### PR TITLE
[flang][driver] rename flang-new to flang

### PR DESCRIPTION
**Pull request by everythingfunctional**
This does a global rename from `flang-new` to `flang`. I also removed/changed any TODOs that I found related to making this change.

### CODE REVIEW

**Comment by kawashima-fj** [comment]
Thanks for the PR.
Is there a symbolic link `flang-new` -> `flang`? We would like to have a transitional period.

**Comment by mjklemm** [comment]
I have seen some more matches for flang-new in the tree:

```
1:.github/workflows/release-binaries.yml:331:          ninja -C ${{ steps.setup-stage.outputs.build-prefix }}/build/tools/clang/stage2-bins/ -j2 flang-new bbc
2:openmp/CMakeLists.txt:72:    set(OPENMP_TEST_Fortran_COMPILER ${LLVM_RUNTIME_OUTPUT_INTDIR}/flang-new)
3:openmp/CMakeLists.txt:74:    set(OPENMP_TEST_Fortran_COMPILER ${LLVM_RUNTIME_OUTPUT_INTDIR}/flang-new.exe)
```

Since I'm travelling, I haven't dug any further at this point.

**Comment by DavidSpickett** [comment]
> Is there a symbolic link flang-new -> flang? We would like to have a transitional period.

Compiler Explorer would also appreciate this. Though I should be able to get it switched to the new name within a week or two.

**Comment by DavidSpickett** [comment]
Also buildbots reference flang-new. We can change the config files very quickly, but getting it into production will be a week or two. A symlink would help here.

**Comment by h-vetinari** [comment]
If you want to compare notes, I recently rebased #74377, and tried to pick up all additional occurrences I could find. Here's the [branch](https://github.com/h-vetinari/llvm-project/tree/flang_rename).

**Comment by everythingfunctional** [comment]
> I have seen some more matches for flang-new in the tree:

Good catch. I'm pretty sure I got them all now

> Is there a symbolic link `flang-new` -> `flang`? We would like to have a transitional period.

I wasn't quite sure it was warranted, but seems like it is so I've added it back.

> If you want to compare notes, I recently rebased #74377,

This was helpful for putting the symlink in, so I gave you co-authorship on that commit. Let me know if you don't actually want it.

**Comment by h-vetinari** [comment]
> This was helpful for putting the symlink in, so I gave you co-authorship on that commit. Let me know if you don't actually want it.

That was all @banach-space's work, I only rebased his commit. Don't think that's worth co-authorship; though ultimately you can decide, I don't mind either way.

**Comment by mjklemm** [comment]
Since we are making this change now, should this PR be updated to follow clang's scheme of having `clang` point to `clang-$version`?

**Comment by banach-space** [comment]
Having multiple active PR's for one change is IMHO rather confusing. It creates unnecessary duplication - why do we expect people to review the same change twice? As a case in point:

> Since we are making this change now, should this PR be updated to follow clang's scheme of having clang point to clang-$version?

That was already included in #74377.

It is absolutely secondary who commits this change (just acknowledging that there are two implementations). However, please make sure that:
* there is a well documented transition period during which `flang-new` is still available as a sym-link (please define _when_ such transition period would end),
* we use this as an opportunity to improve Flang's consistency with Clang (please introduce `flang-$version`),
* this change does not break Flang's buildbots and all customers are well supported throughout the renaming (there might be a stream of bug reports following this),
* community members who contributed to the previous PRs for this (either through code, reviews or testing) are aware that the discussion has moved here and all changes have been attributed accordingly.

Thank you all for working on this! 🙏🏻

**Comment by everythingfunctional** [comment]
> Since we are making this change now, should this PR be updated to follow clang's scheme of having `clang` point to `clang-$version`?

Sure. I've added what I believe does that.

> Having multiple active PR's for one change is IMHO rather confusing. It creates unnecessary duplication - why do we expect people to review the same change twice?

Apologies. I didn't mean to create a duplicate or hijack this, but I thought any previous PRs were likely stale and would be more difficult to resurrect than to just rerun a simple `find`/`sed` command.

**Comment by kiranchandramohan** [comment]
Could you add an experimental message ("The openmp support in Flang is experimental") when compiling with OpenMP? The warning can be in the Driver code that forwards the `-fopenmp` flag to the driver. https://github.com/llvm/llvm-project/blob/47861fa3f557b8017529de530914065034dbcde7/clang/lib/Driver/ToolChains/Flang.cpp#L780
If you are not familiar, you can search for `diag::warn` in the Driver code to understand how to add driver warnings.

Note: We can remove the experimental message whenever the items in the following meta-ticket is complete. We can talk about the contents of the ticket separately.
https://github.com/llvm/llvm-project/issues/110008

**Comment by mjklemm** [comment]
> Could you add an experimental message ("The openmp support in Flang is experimental") when compiling with OpenMP? The warning can be in the Driver code that forwards the `-fopenmp` flag to the driver.

Hm.  I'm not yet fully convinced that we should be doing this.  Shouldn't `-fopenacc` then also emit such a warning?

**Comment by everythingfunctional** [comment]
> Shouldn't `-fopenacc` then also emit such a warning?

Well, we could get into a philosophical discussion about whether software is ever *not* experimental. But in this case I think it's fine to be a bit proactive and warn users that we know there are still some unfinished aspects. Or we could maybe use the word "incomplete" instead of experimental. Either way I'm happy to add the warning for both, just let me know.

**Comment by clementval** [comment]
> > Could you add an experimental message ("The openmp support in Flang is experimental") when compiling with OpenMP? The warning can be in the Driver code that forwards the `-fopenmp` flag to the driver.
> 
> Hm. I'm not yet fully convinced that we should be doing this. Shouldn't `-fopenacc` then also emit such a warning?

OpenACC has no codegen upstream yet so it will never produce an executable. When we start adding support for codeine/runtime, a warning is probably nice until good enough support is reached.

**Comment by everythingfunctional** [comment]
Sounds like I should just add the warning to OpenMP for now. I'll work on getting that added and then figure out why the CI is failing.

File: flang/docs/FlangDriver.md
Comment (luporl) [inline flang/docs/FlangDriver.md:338]:
```suggestion
(and CMake 3.24.0), `cmake` can detect `flang` as a
```
Can CMake already detect `flang` as well?

Comment (AlexisPerry) [reply flang/docs/FlangDriver.md:338]: This is a good question.  Also, can it detect our flang instead of classic Flang regardless of executable name, or is the detection of flang vs. flang-new the way CMake differentiates the two compilers?

Comment (everythingfunctional) [reply flang/docs/FlangDriver.md:338]: That I am not quite sure of. I will look into it and adjust that comment in the docs as appropriate.

Comment (everythingfunctional) [reply flang/docs/FlangDriver.md:338]: @banach-space appears to have been involved in that MR, do you happen to know the answer to the above questions? It looks like it doesn't rely on the `flang-new` name, but I'm not sure.

Comment (hakostra) [reply flang/docs/FlangDriver.md:338]: 1. A search in the CMake repository revel that `flang-new` only appear in testcases and ci-related code - not in any code that eventually end up in CMake. This indicate that the name `flang-new` is not a magic token for CMake to recognize LLVM Flang.
2. I think that [this is the merge](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/7246) that first introduced LLVM Flang to CMake. It seems like the check is through preprocessor macros defined by the different compilers. See [`Modules/CMakeFortranCompilerId.F.in`](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/CMakeFortranCompilerId.F.in) in the CMake sources.

Comment (banach-space) [reply flang/docs/FlangDriver.md:338]: >  It seems like the check is through preprocessor macros defined by the different compilers.

Yes, thank you @hakostra 🙏🏻

Comment (DavidTruby) [reply flang/docs/FlangDriver.md:338]: Just to add, for an actually working CMake with flang-new that detects the correct options and does the right thing reliably, you'll need CMake 3.28 as that's the first version with fully working flang-new support, so if updating this comment could you change 3.24 to 3.28 as well? 
I am fairly sure the CMake support for LLVM flang is designed to work fine regardless of the name of the binary, as the CMake devs were anticipating this name change anyway.

File: clang/lib/Driver/ToolChains/Flang.cpp
Comment (tblah) [inline clang/lib/Driver/ToolChains/Flang.cpp:792]: Please could you add a test that this warning is printed. `flang/test/Driver/fopoenmp.f90` would be a good place. There is an example in the file already checking for a warning.

File: flang/docs/FlangDriver.md
Comment (everythingfunctional) [reply flang/docs/FlangDriver.md]: So, sounds like support is there in CMake, so I've updated the note and the identified version.

File: clang/lib/Driver/ToolChains/Flang.cpp
Comment (everythingfunctional) [reply clang/lib/Driver/ToolChains/Flang.cpp:792]: While I was slightly hesitant to add a test for a temporary warning, I've done so. If anyone else thinks it doesn't belong I'll undo it.

File: flang/docs/FlangDriver.md
Comment (banach-space) [inline flang/docs/FlangDriver.md:338]: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/7246 pre-dates CMake 3.28.0 by quite a few releases, I would just remove the link.

@DavidTruby what makes 3.28 special? Should 3.28 be marked as requirement?

**Review by kiranchandramohan — APPROVED** [review]
LGTM. Thanks for adding the warning about OpenMP.

Please ask in the Flang community call tomorrow and in the discourse post before you submit.

**Comment by everythingfunctional** [comment]
@kiranchandramohan , I don't have commit access, so could you merge for me?

**Comment by h-vetinari** [comment]
Congratulations on this huge milestone to all involved in flang! 🍾 🥳 🚀

**Comment by RichBarton-Arm** [comment]
+1000
Great to see this finally happen. Great job everyone!

**Comment by klausler** [comment]
There are multiple weird failures in llvm-test-suite where clang++ complains about unknown options that look like they're slang-only options.  Did you run llvm-test-suite before merging?  Please check.  These failures are going to cause build bot failures shortly in the clang-*-2stage builds if not repaired or reverted.

**Comment by everythingfunctional** [comment]
@klausler , can you point me specifically at the failure messages? The failures I've seen so far from the merge did not look related to my (admittedly untrained) eye. I ran the tests in the repo and they passed locally. Are there other tests somewhere that need to be updated as well?

**Comment by kiranchandramohan** [comment]
The 2 stage buildbot (https://lab.llvm.org/buildbot/#/builders/41) with the Fortran tests passed after this merge. Its latest run is also a pass.

**Comment by kiranchandramohan** [comment]
@everythingfunctional Following are the steps to run the testsuite. Please check and confirm if you did not run it before.

```
git clone https://github.com/llvm/llvm-test-suite.git
cd llvm-test-suite
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=$HOME/llvm-project/build_release/bin/clang  -DCMAKE_CXX_COMPILER=$HOME/llvm-project/build_release/bin/clang++ -DCMAKE_Fortran_COMPILER=$HOME/llvm-project/build_release/bin/flang-new -DTEST_SUITE_FORTRAN=On -DTEST_SUITE_SUBDIRS=Fortran -DTEST_SUITE_FORTRAN_ISO_C_HEADER_DIR=$HOME/llvm-project/flang/include/flang ../
make -j48
NO_STOP_MESSAGE=1 $HOME/llvm-project/build_release/bin/llvm-lit -v  .
```

**Comment by everythingfunctional** [comment]
> @everythingfunctional Following are the steps to run the testsuite. Please check and confirm if you did not run it before.
> 
> ```
> git clone https://github.com/llvm/llvm-test-suite.git
> cd llvm-test-suite
> mkdir build
> cd build
> cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=$HOME/llvm-project/build_release/bin/clang  -DCMAKE_CXX_COMPILER=$HOME/llvm-project/build_release/bin/clang++ -DCMAKE_Fortran_COMPILER=$HOME/llvm-project/build_release/bin/flang-new -DTEST_SUITE_FORTRAN=On -DTEST_SUITE_SUBDIRS=Fortran -DTEST_SUITE_FORTRAN_ISO_C_HEADER_DIR=$HOME/llvm-project/flang/include/flang ../
> make -j48
> NO_STOP_MESSAGE=1 $HOME/llvm-project/build_release/bin/llvm-lit -v  .
> ```

Only failure I got was:

```text
Failed Tests (1):
  test-suite :: Fortran/gfortran/regression/gomp/gfortran-regression-compile-regression__gomp__proc_ptr_1_f90.test
```

Which I don't think the rename is at fault for.

**Comment by sylvestre** [comment]
@everythingfunctional with this change and install, we are installing flang with the -20 in the install dir:
` -- Installing: /build/source/debian/tmp/usr/lib/llvm-20/bin/flang-20 `

it should not be the case, it should be named `/usr/lib/llvm-20/bin/flang` to match what is done elsewhere

**Comment by everythingfunctional** [comment]
> @everythingfunctional with this change and install, we are installing flang with the -20 in the install dir: `-- Installing: /build/source/debian/tmp/usr/lib/llvm-20/bin/flang-20`
> 
> it should not be the case, it should be named `/usr/lib/llvm-20/bin/flang` to match what is done elsewhere

Forgive me if I misunderstood or implemented this incorrectly, but I was under the impression (based on [an earlier comment](https://github.com/llvm/llvm-project/pull/110023#issuecomment-2378556647)) that this was the intended behavior.

**Comment by hakostra** [comment]
I have freshly compiled and installed LLVM flang after the rename. I have:

```
$ ll flang*
lrwxrwxrwx 1 hakostra hakostra         8 okt.  15 17:03 flang -> flang-20*
-rwxr-xr-x 1 hakostra hakostra 176517648 okt.  23 12:03 flang-20*
lrwxrwxrwx 1 hakostra hakostra         5 okt.  23 12:13 flang-new -> flang*
$ ll clang*
lrwxrwxrwx 1 hakostra hakostra         8 okt.  15 17:03 clang -> clang-20*
lrwxrwxrwx 1 hakostra hakostra         5 okt.  23 12:13 clang++ -> clang*
-rwxr-xr-x 1 hakostra hakostra 156923176 okt.  23 11:49 clang-20*
```

so `flang-20` and `clang-20` are the actual binaries, `flang` and `clang` are symlinks to these. Also, on my system, `/usr/bin/gcc` is just a symlink to whatever `gcc` version I have, for in my case `/usr/bin/gcc-13`. Same for `gfortran`. So it's the same naming scheme.

This is intended behavior. If `flang` was the name of the binary and `flang-20` the symlink, it would not be possible to have several versions installed in besides each other.

**Comment by sylvestre** [comment]
Did you try with a prefix and then ninja install?

It has a different behavior between Clang and flang

**Comment by hakostra** [comment]
I use `-DCMAKE_INSTALL_PREFIX=` as an argument to `cmake` and `ninja install` - that gives the layout above.

**Comment by sylvestre** [comment]
OK, bizarre.
With apt.llvm.org, flang is the only one with the prefix (-20) in /usr/lib/llvm-20/bin/

**Pull request merged**
